src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php

```diff
@@ -0,0 +1,590 @@
+            SELECT DISTINCT cp.id_product, c.id_cart, c.id_shop, cp.id_shop AS cp_id_shop
+            FROM ' . _DB_PREFIX_ . 'cart_product cp
+            JOIN ' . _DB_PREFIX_ . 'cart c ON (c.id_cart = cp.id_cart)
+            JOIN ' . _DB_PREFIX_ . 'product p ON (cp.id_product = p.id_product)
+            WHERE c.id_customer = ' . (int) $customer->id . '
+                AND NOT EXISTS (
+                        SELECT 1
+                        FROM ' . _DB_PREFIX_ . 'orders o
+                        JOIN ' . _DB_PREFIX_ . 'order_detail od ON (o.id_order = od.id_order)
+                        WHERE product_id = cp.id_product AND o.valid = 1 AND o.id_customer = ' . (int) $customer->id . '
+                )
+        ';
+
+        $viewedProductsData = Db::getInstance()->executeS($sql);
+        foreach ($viewedProductsData as $productData) {
+            $product = new Product(
+                $productData['id_product'],
+                false,
+                $this->contextLangId,
+                $productData['id_shop']
+            );
+
+            if (!Validate::isLoadedObject($product)) {
+                continue;
+            }
+
+            $productUrl = $this->link->getProductLink(
+                $product->id,
+                $product->link_rewrite,
+                Category::getLinkRewrite($product->id_category_default, $this->contextLangId),
+                null,
+                null,
+                $productData['cp_id_shop']
+            );
+
+            $viewedProducts[] = new ViewedProductInformation(
+                (int) $product->id,
+                Tools::htmlentitiesUTF8($product->name),
+                $productUrl
+            );
+        }
+
+        return new ProductsInformation(
+            $boughtProducts,
+            $viewedProducts
+        );
+    }
+
+    /**
+     * @param Customer $customer
+     *
+     * @return MessageInformation[]
+     */
+    private function getCustomerMessages(Customer $customer)
+    {
+        $customerMessages = [];
+        $messages = CustomerThread::getCustomerMessages((int) $customer->id);
+
+        $messageStatues = [
+            'open' => $this->translator->trans('Open', [], 'Admin.Orderscustomers.Feature'),
+            'closed' => $this->translator->trans('Closed', [], 'Admin.Orderscustomers.Feature'),
+            'pending1' => $this->translator->trans('Pending 1', [], 'Admin.Orderscustomers.Feature'),
+            'pending2' => $this->translator->trans('Pending 2', [], 'Admin.Orderscustomers.Feature'),
+        ];
+
+        foreach ($messages as $message) {
+            $status = isset($messageStatues[$message['status']]) ?
+                $messageStatues[$message['status']] :
+                $message['status']
+            ;
+
+            $customerMessages[] = new MessageInformation(
+                (int) $message['id_customer_thread'],
+                substr(strip_tags(html_entity_decode($message['message'], ENT_NOQUOTES, 'UTF-8')), 0, 75),
```

Comment:
😅

Reply:
i think this can be done in twig? :D

Reply:
Is this processing intended to *decode* the stored data or is it intended to *display properly* the data in an HTML page ?
If answer is 1, we keep it here. If answer is 2, this goes in the template

Reply:
to be honest, i dont really know why it was done this way. :/ i've simplified it a little.